### PR TITLE
ASoC: Intel: sof-rt5682: minor fixes in AMP selection and codec button

### PR DESCRIPTION
**Pull request by sathyap-chrome**
…utton

This fixes 2 issues listed when using with platform using speaker Amp.
1. The selection order of m/c in match table.
2. BIT mask for AMP SSP was incorrect.

Other fixes
1. The RT5682 codec button mapping is corrected.

Change-Id: Icd714ca2d21a17ffa25a40c524b7c387988b8d8f
Signed-off-by: Sathya Prakash M R <sathya.prakash.m.r@intel.com>

### CODE REVIEW

File: sound/soc/intel/boards/sof_rt5682.c [hunk 1]
```diff
@@ -29,9 +29,9 @@
 #define SOF_RT5682_MCLK_EN			BIT(3)
 #define SOF_RT5682_MCLK_24MHZ			BIT(4)
 #define SOF_SPEAKER_AMP_PRESENT		BIT(5)
-#define SOF_RT5682_SSP_AMP(quirk)		((quirk) & GENMASK(8, 6))
-#define SOF_RT5682_SSP_AMP_MASK			(GENMASK(8, 6))
 #define SOF_RT5682_SSP_AMP_SHIFT		6
+#define SOF_RT5682_SSP_AMP(quirk)		(((quirk) << SOF_RT5682_SSP_AMP_SHIFT) & GENMASK(8, 6))
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_rt5682.c]: That looks just wrong. You shift by 6 bits and then mask with bits 8..6? WTH?

File: sound/soc/intel/boards/sof_rt5682.c [hunk 2]
```diff
@@ -144,9 +144,9 @@ static int sof_rt5682_codec_init(struct snd_soc_pcm_runtime *rtd)
 	jack = &ctx->sof_headset;
 
 	snd_jack_set_key(jack->jack, SND_JACK_BTN_0, KEY_PLAYPAUSE);
-	snd_jack_set_key(jack->jack, SND_JACK_BTN_1, KEY_VOLUMEUP);
-	snd_jack_set_key(jack->jack, SND_JACK_BTN_2, KEY_VOLUMEDOWN);
-	snd_jack_set_key(jack->jack, SND_JACK_BTN_3, KEY_VOICECOMMAND);
+	snd_jack_set_key(jack->jack, SND_JACK_BTN_1, KEY_VOICECOMMAND);
+	snd_jack_set_key(jack->jack, SND_JACK_BTN_2, KEY_VOLUMEUP);
+	snd_jack_set_key(jack->jack, SND_JACK_BTN_3, KEY_VOLUMEDOWN);
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_rt5682.c:150]: can you double check please, there are 3 versions on the mapping in the upstream drivers.

File: sound/soc/intel/common/soc-acpi-intel-cnl-match.c [hunk 3]
```diff
@@ -37,17 +37,17 @@ struct snd_soc_acpi_mach snd_soc_acpi_intel_cnl_machines[] = {
...
+		.sof_tplg_filename = "sof-cml-rt5682-max98357a.tplg",
 	},
 	{
-		.id = "MX98357A",
+		.id = "10EC5682",
 		.drv_name = "sof_rt5682",
-		.quirk_data = &cml_codecs,
 		.sof_fw_filename = "sof-cnl.ri",
-		.sof_tplg_filename = "sof-cml-rt5682-max98357a.tplg",
+		.sof_tplg_filename = "sof-cml-rt5682.tplg",
```
Comment (plbossart) [inline sound/soc/intel/common/soc-acpi-intel-cnl-match.c:42]: this should be a separate patch please.

File: sound/soc/intel/boards/sof_rt5682.c (hunk 1 above)
Comment (sathyap-chrome) [reply sound/soc/intel/boards/sof_rt5682.c]: we are using single bitmask for all the different representation : CODEC, CODEC SSP port identification, CLK SRC, from 6th bit i use for Speaker.
if SSP of AMP is 1, we do 1 << 6 and then mask to ensure end result is 1.

Let me relook again with the math. I did few calculations.

File: sound/soc/intel/boards/sof_rt5682.c (hunk 2 above)
Comment (sathyap-chrome) [reply sound/soc/intel/boards/sof_rt5682.c:150]: This is verified by @naveen on the CML platform and he suggested these changes. @bardliao can you confirm on RVP if you see same mapping ?

File: sound/soc/intel/common/soc-acpi-intel-cnl-match.c (hunk 3 above)
Comment (sathyap-chrome) [reply sound/soc/intel/common/soc-acpi-intel-cnl-match.c:42]: Sure. will do

File: sound/soc/intel/boards/sof_rt5682.c (hunk 1 above)
Comment (ranj063) [reply sound/soc/intel/boards/sof_rt5682.c]: @sathyap-chrome I added the SOF_RT5682_SSP_AMP_SHIFT based on the GENMASK(8, 6) for the SSP_AMP(quirk). So what you are doing here is incorrect.

I guess you need to verify what the GEN_MASK for the AMP quirk is then fix that and the shift accordingly.

Comment (sathyap-chrome) [reply sound/soc/intel/boards/sof_rt5682.c]: @ranj063  Here are prints from my hatch.
ERR kernel: [    6.962845]  lets Prints all values here:
ERR kernel: [    6.962850]  This is hatch - so SSP AMP should be 1 and ssp_codec should be 0
ERR kernel: [    6.962854] sof_rt5682 sof_rt5682: sof_rt5682_quirk = **0x78**
 [Sathya: 1111000 (bit 3,bit 4 both 1. Bit 5 =1 (spk present) BIT 6 =1 (SSP of SPK)
ERR kernel: [    6.962856]  GENMASK (8,6) = 0x1c0
ERR kernel: [    6.962858]  Earlier way of AMP MASK :
ERR kernel: [    6.962860]  SOF_RT5682_SSP_AMP (1 & GENMASK) = 0x0 
[Sathya : this results in SSP of AMP being returned as 0]
ERR kernel: [    6.962861]  Current way of AMP MASK :
ERR kernel: [    6.962863]  SOF_RT5682_SSP_AMP (1 <<6 & Then GENMASK) = 0x40 
[ Sathya : 1000000 [ Bit 6 = 1 ] Now we right shift and we get SSP as 1]

My understanding of the current implementation:
BIT 0 - 2   : used for identifying the SSP of the CODEC
BIT 3        : check on MCLK
BIT 4        : use 24 Mhz MCLK
BIT 5        : is AMP Present
BIT 6-8     : SSP of the AMP 

So the sof_rt5682_quirk should be " **0x78**"  for CML platform with HS codec on SSP0 and SPK AMP on SSP1.

So when setting BIT, we left shift ( (quirk) << SOF_RT5682_SSP_AMP_SHIFT)and then when reading back we right shift. ( (sof_rt5682_quirk & SOF_RT5682_SSP_AMP_MASK) >>
                        SOF_RT5682_SSP_AMP_SHIFT;)



below code snippet for reference: 
+       printk(KERN_ERR " lets Prints all values here:\n");
+       printk(KERN_ERR " This is hatch - so SSP AMP should be 1 and ssp_codec should be 0\n");
+       dev_err(&pdev->dev, "sof_rt5682_quirk = 0x%lx\n", sof_rt5682_quirk);
+
+       printk(KERN_ERR " GENMASK (8,6) = 0x%lx \n", GENMASK (8,6));
+       printk(KERN_ERR " Earlier way of AMP MASK :\n");
+       printk(KERN_ERR " SOF_RT5682_SSP_AMP (1 & GENMASK) = 0x%lx\n", 1 & GENMASK (8,6));
+       printk(KERN_ERR " Current way of AMP MASK : \n");
+       printk(KERN_ERR " SOF_RT5682_SSP_AMP (1 <<6 & Then GENMASK) = 0x%lx\n", (1<<6) & GENMASK (8,6));

Comment (sathyap-chrome) [reply sound/soc/intel/boards/sof_rt5682.c]: ok. i guess i found the fix. let me post the v2. the GENMASK in the 1st place is not needed. That should be right way.

Comment (bardliao) [reply sound/soc/intel/boards/sof_rt5682.c]: This looks good to me. BTW,  it seems the GENMASK is not needed on ``#define SOF_RT5682_SSP_CODEC(quirk)		((quirk) & GENMASK(2, 0))``, too.

File: sound/soc/intel/boards/sof_rt5682.c (hunk 2 above)
Comment (bardliao) [reply sound/soc/intel/boards/sof_rt5682.c:150]: @sathyap-chrome Sorry, I am not able to check it on RVP since the IRQ is still not working on RVP. I think we should verify it on a real product.

Comment (sathyap-chrome) [reply sound/soc/intel/boards/sof_rt5682.c:150]: Thanks bard. i think we are good. I'll ask naveen/yong to check on GLK too  if button mapping is as corrected

File: sound/soc/intel/boards/sof_rt5682.c (hunk 1 above)
Comment (sathyap-chrome) [reply sound/soc/intel/boards/sof_rt5682.c]: true. But i think we are good with v2 now. Let @ranj063 and Pierre approve :)

File: sound/soc/intel/boards/sof_rt5682.c [hunk 4]
```diff
@@ -29,9 +29,9 @@
 #define SOF_RT5682_MCLK_EN			BIT(3)
 #define SOF_RT5682_MCLK_24MHZ			BIT(4)
 #define SOF_SPEAKER_AMP_PRESENT		BIT(5)
-#define SOF_RT5682_SSP_AMP(quirk)		((quirk) & GENMASK(8, 6))
-#define SOF_RT5682_SSP_AMP_MASK			(GENMASK(8, 6))
 #define SOF_RT5682_SSP_AMP_SHIFT		6
+#define SOF_RT5682_SSP_AMP(quirk)		((quirk) << SOF_RT5682_SSP_AMP_SHIFT)
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_rt5682.c]: this makes no sense. You are passing a 1 as a parameter, so this can be replaced by SOF_SPEAKER_AMP_PRESENT

You don't need these 3 definitions line 32..34 at all.

Comment (sathyap-chrome) [reply sound/soc/intel/boards/sof_rt5682.c]: actually the AMP_PRESENT only tells if AMP is present. The quirk is to tell the SSP its connected to.
We can have a platform where AMP is present and can be on SSP2 or maybe SSP0 as well.

Hence have separate definitions.

Comment (plbossart) [reply sound/soc/intel/boards/sof_rt5682.c]: ok, then you do need a mask to protect the quirk.
It should be 

define SOF_RT5682_SSP_AMP_MASK			(GENMASK(8, 6)
define SOF_RT5682_SSP_AMP(quirk)		(((quirk) << SOF_RT5682_SSP_AMP_SHIFT) & SOF_RT5682_SSP_AMP_MASK)

and you use it as 

ssp_amp = (sof_rt5682_quirk & SOF_RT5682_SSP_AMP_MASK) >>			s
			SOF_RT5682_SSP_AMP_SHIFT;

Comment (sathyap-chrome) [reply sound/soc/intel/boards/sof_rt5682.c]: This seems to be same as 1st way i did in v1 ? should i bring back that way as in v1 then

Comment (plbossart) [reply sound/soc/intel/boards/sof_rt5682.c]: The problem is that you use quirk as a parameter when we also have a sof_rt5682_quirk variable, so this is confusing. You are not shifting the quirk, you have shifting a configuration to fit it into the quirk.

Comment (sathyap-chrome) [reply sound/soc/intel/boards/sof_rt5682.c]: @pierre, i did the way you suggested in above comments and submitted the patch again.

File: sound/soc/intel/boards/sof_rt5682.c (hunk 2 above)
Comment (plbossart) [reply sound/soc/intel/boards/sof_rt5682.c:150]: For da7219, the mapping is
 	snd_jack_set_key(jack->jack, SND_JACK_BTN_0, KEY_PLAYPAUSE);
	snd_jack_set_key(jack->jack, SND_JACK_BTN_1, KEY_VOLUMEUP);
	snd_jack_set_key(jack->jack, SND_JACK_BTN_2, KEY_VOLUMEDOWN);
	snd_jack_set_key(jack->jack, SND_JACK_BTN_3, KEY_VOICECOMMAND);

Please double-check.

File: sound/soc/intel/boards/sof_rt5682.c [hunk 5]
```diff
@@ -519,6 +520,7 @@ static int sof_audio_probe(struct platform_device *pdev)
 
 	/* compute number of dai links */
 	sof_audio_card_rt5682.num_links = 1 + dmic_num + hdmi_num;
+
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_rt5682.c:523]: remove newline

File: sound/soc/intel/boards/sof_rt5682.c (hunk 2 above)
Comment (naveen-manohar) [reply sound/soc/intel/boards/sof_rt5682.c:150]: For DA7219 its correct,
Below is the correct button mapping for rt5682 codec, GLK variant of this codec too has the same mapping.
	snd_jack_set_key(jack->jack, SND_JACK_BTN_0, KEY_PLAYPAUSE);
	snd_jack_set_key(jack->jack, SND_JACK_BTN_1, KEY_VOICECOMMAND);
	snd_jack_set_key(jack->jack, SND_JACK_BTN_2, KEY_VOLUMEUP);
	snd_jack_set_key(jack->jack, SND_JACK_BTN_3, KEY_VOLUMEDOWN);

File: sound/soc/intel/boards/sof_rt5682.c (hunk 5 above)
Comment (bardliao) [reply sound/soc/intel/boards/sof_rt5682.c:523]: @sathyap-chrome Did you add the newline intentionally? If not, please remove it. Thanks

**Comment by sathyap-chrome** [comment]
I just checked. I added newline to have a line break before if for better readability. let me know if its OK to have the same.
